docs/pipe-protocol.md

```diff
@@ -0,0 +1,270 @@
+
+Handshake is expected to happen in all modes. Help, discovery, and execution.
+Today, there is an MTP bug that we don't handshake in help mode. This will be worked around in .NET CLI. A fix needs to be done in MTP.
+
+### Handshake request
+
+The handshake request contains multiple properties. Some of which are required by the .NET CLI, and some of them are not required.
+
+- Process ID (optional): The process id sending the handshake.
+- Architecture (required): The architecture of the running test application.
+- Framework (required): The framework description, as given by `RuntimeInformation.FrameworkDescription`.
+- OS (optional): The operating system running the test application.
+- Supported protocol versions (required): The protocol versions that are supported by the test application, separated by semicolons.
+- Host type (required): The host type which is handshaking with us.
+- Module path (required): The path to the test module, either the assembly dll path, or the actual apphost (exe) path.
+- ExecutionId (required): Explained above.
+- InstanceId (required): Explained above.
+
+### Handshake response
+
+The handshake response contains multiple properties. Some of which are required by the test application, and some of them are not required.
+
+- Process ID (optional): The process id of the .NET CLI.
+- Architecture (optional): The architecture of the .NET CLI.
+  - TODO(youssef): This shouldn't be needed at all.
+- Framework (optional): The framework description of the .NET CLI, as given by `RuntimeInformation.FrameworkDescription`
+  - TODO(youssef): This shouldn't be needed at all.
+- OS (optional): The operating system running the .NET CLI.
+- Supported protocol version: The final protocol version to use. This can be empty/omitted if the .NET CLI doesn't support any of the versions sent during the handshake request.
+  - TODO(youssef): How should the version checks happen? Should we consider only the "major" part for deciding compatibility? In what scenarios could we need to only bump minor/patch components? What would it signify and how that info is supposed to be used? Today, we do a full exact version check.
+
+## Test session event message
+
+TODO(Youssef): Verify if this should be sent during help or discovery modes.
+
+This message denotes test events. Currently, there are two events, start and end. This message has the following properties:
+
+- Event type (required): start or end.
+- SessionUid (required)
+- ExecutionId (required)
+
+## Command line options
+
+This message is expected to be received only in help mode.
```

Comment:
it would be useful to explain what this message means from user's point of view, i.e. that if a user calls `dotnet test --help` each test application is expected to report its command line options back